site/src/api/queries/templates.ts

```diff
@@ -133,3 +120,53 @@ export const templateVersionExternalAuth = (versionId: string) => {
     queryFn: () => API.getTemplateVersionExternalAuth(versionId),
   };
 };
+
+export const createTemplate = () => {
+  return {
+    mutationFn: createTemplateFn,
+  };
+};
+
+const createTemplateFn = async (options: {
+  organizationId: string;
+  version: CreateTemplateVersionRequest;
+  template: Omit<CreateTemplateRequest, "template_version_id">;
+}) => {
+  const version = await API.createTemplateVersion(
+    options.organizationId,
+    options.version,
+  );
+  await waitBuildToBeFinished(version);
+  return API.createTemplate(options.organizationId, {
+    ...options.template,
+    template_version_id: version.id,
+  });
+};
+
+const waitBuildToBeFinished = async (version: TemplateVersion) => {
+  let data: TemplateVersion;
+  let jobStatus: ProvisionerJobStatus;
+  do {
+    await delay(1000);
+    data = await API.getTemplateVersion(version.id);
+    jobStatus = data.job.status;
+
+    if (jobStatus === "succeeded") {
+      return version.id;
+    }
+  } while (jobStatus === "pending" || jobStatus === "running");
+
+  // No longer pending/running, but didn't succeed
+  throw new JobError(data.job, version);
+};
+
+export class JobError extends Error {
+  public job: ProvisionerJob;
+  public version: TemplateVersion;
+
+  constructor(job: ProvisionerJob, version: TemplateVersion) {
+    super(job.error);
+    this.job = job;
+    this.version = version;
+  }
+}
```

Comment:
do we have somewhere more general we can put these sorts of errors? not crazy about this being tucked away in queries/

Reply:
We don't but I'm open to it if you have any place or organization in mind.

Reply:
maybe util/errors or something? I don’t tend to like util/ folders, but it’d be slightly more reusable there until we think of somewhere better

Reply:
Hm... Idk but I feel it is kinda of far from where it is used since it is only used there - for now and probably it will be the only place where it makes sense IMO.